ansible_base/lib/utils/address.py

```diff
@@ -0,0 +1,125 @@
+import enum
+import ipaddress
+import re
+
+
+class AddressType(enum.Enum):
+    """
+    AddressType provides an abstracted identification of the determined kind of
+    an address.  The abstraction eliminates cohesion between the provided
+    address identification functionality and its clients.
+
+    The type allows a client which may be configured with any of the supported
+    types to identify which type was specified and perform necessary runtime
+    handling.
+
+    An example would be a client which uses the values from its configuration
+    to construct URLs and which is configured using raw IPv6 addreesses. The
+    client needs to be able to determine the address type to know what
+    processing (such as enclosing IPv6 addresses in []s) is needed to
+    successfully utilize it.
+    """
+
+    HOSTNAME = "hostname"
+    IPv4 = "ipv4"
+    IPv6 = "ipv6"
+    UNKNOWN = "unknown"
+
+
+def _classify_address_string(address_string) -> AddressType:
+    """
+    Categorizes a given string as IPv4, IPv6, hostname, or unknown.
+
+    Args:
+        address_string: The string to categorize.
+
+    Returns:
+        A value of AddressType indicating the category.
+    """
+    try:
+        ipaddress.IPv4Address(address_string)
+        return AddressType.IPv4
+    except ipaddress.AddressValueError:
+        pass
+
+    try:
+        ipaddress.IPv6Address(address_string)
+        return AddressType.IPv6
+    except ipaddress.AddressValueError:
+        pass
+
+    # Basic hostname check (can be expanded for more rigorous validation)
+    if re.match(r"^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$", address_string):
+        return AddressType.HOSTNAME
+
+    return AddressType.UNKNOWN
+
+
+def classify_address_string(address_string) -> AddressType:
+    """
+    Categorizes a given string as IPv4, IPv6, hostname, or unknown.
+
+    An IPv6 address wrapped in [] is considered a hostname.
+
+    Args:
+        address_string: The string to categorize.
+
+    Returns:
+        A value of AddressType indicating the category.
+    """
+    address_type = _classify_address_string(address_string)
+    if address_type != AddressType.UNKNOWN:
+        return address_type
+
+    # We could be dealing with an IPv6 address wrapped in [].
+    # If that is the case we want to identify it as a host name.
+    if (address_string[0] == "[") and (address_string[-1] == "]"):
+        address_type = _classify_address_string(address_string[1:-1])
+
+        # Only an address type of IPv6 is considered valid here.
+        # We don't want to treat any other specification as valid.
+        if address_type == AddressType.IPv6:
+            return AddressType.HOSTNAME
```

Comment:
Is HOSTNAME correct here? Wouldn't it be an IPv6 or do we consider [<ip v6 address>] as a hostname?

Reply:
The comment before the preceding if can be more explicit.  The intention here is two-fold:
1. minimize the type space; i.e., not have two kinds of IPv6
2. minimize the logic that a client has to implement 

In the case of 2 the way in which a []ed IPv6 is used is as if it were a hostname.  That is, no special processing is required (beyond convenience this is an additional reason there's a method that does the splitting into "address" and "port"; otherwise the client would still have to special case processing). 

So the only, if it matters to the component, address type it may have to care about is the single IPv6 type.

Reply:
Changed to categorize IPv6, wrapped or not, as IPv6.  Added a property to the AddressTypeResponse which will return the bracketed IPv6 or None if the address type is not IPv6.